### ISSUE TITLE
"Pause For..." timers from Stopped state often missed

### ISSUE THREAD

**Issue by ferdnyc** [BiglySoftware/BiglyBT#1054]
Using the "Advanced > Pause For..." feature to create more than one active timer on a stopped torrent appears to frequently cause _all_ of the timers to  be missed, leaving the torrents in Paused state forever, still showing the "PAUSED (RESUMING AT _HH:MM:DD_)" for a time already passed.

### Steps to reproduce
1. Load multiple torrents in the "Stopped" state
2. Right-click the first, choose "Advanced > Pause For...", and enter 1 minute for the timeout.
3. Repeat for the second torrent, repeat again for the third.
4. Wait 2 minutes. Most of the time, none of the three torrents will have started.

It seems to _mostly_* work fine if you limit it to only a **single** timer, even on multiple torrents. But set an additional one before the first one fires, and they **all** break.

\* – (I've had some of those get missed, too. But that was after triggering this bug, so it may have been residual breakage.)

So far, this _appears_ to be limited to torrents coming from the Stopped state. I was able to pause multiple Downloading and Seeding torrents, and they all resumed as scheduled.

The proximity of the expiration times doesn't appear to matter — setting them to fire one right after another, as in the steps above, is just a contrived way of efficiently demonstrating the issue. I initially encountered this when I set a group of ~ 3-4 torrents to start after 25 minutes, then a second group of ~ 5-6 to start in 60 minutes (35 minutes later than the first group). A few hours later I noticed that none of them had started.

While creating this bug report I also had BiglyBT crash on me one time. I selected "Advanced > Pause For..." on a torrent which was showing "PAUSED (RESUMING..." with a past time, and the main window instantly vanished. But that _usually_ works fine, so I'm not sure what the issue was. I can't seem to reproduce that part of it now.

## System
```
Java 1.8.0_212 (64 bit)
  Oracle Corporation
/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.212.b04-0.fc30.x86_64/jre

SWT v4924, gtk/3.24.8, zoom=100, dpi=96
Linux v5.0.9-301.fc30.x86_64, amd64 (64 bit)
B1.9.0.1_B17/4 az3
```

**Comment by parg** [comment]
Hmm, I see a bug with selecting 'pause for' on an already stopped torrent (pausing stopped torrents used to be disallowed but when this was changed I suppose this bug crept into the 'pause for' logic). However, it is a general bug in that these torrents will never resume, not related to how many of them there are established etc.

**Comment by ferdnyc** [comment]
> pausing stopped torrents used to be disallowed but when this was changed

Yeah, that was at my request actually. 😇 #727 .

> However, it is a general bug in that these torrents will never resume, not related to how many of them there are established etc.

Huh, that's interesting. I mean, if it's an intermittent bug, maybe it was just law of averages: I was more likely to hit it when I had multiple different pause expirations, so I convince myself that was one of the conditions for triggering the bug when it was really unrelated.

But, "Pause For..." from a stopped torrent definitely _sometimes_ works, I've used it plenty since that ability was added. I just hadn't tried using it twice within the same period before.

**Comment by parg** [comment]
The bug is that if the download isn't already paused then the 'resume-time' is temporarily stored as a negative value, the download is paused and the state transition from whatever->stopped that follows is used to set the 'resume-time' to the correct positive value. However, in the case of the download already being stopped this transition never occurs and the time is left negative. When the auto-resume timer expires it checks that the download's resume-time is as expected before resuming the download. In the case where it has been left negative this check fails and the resume doesn't occur.

If you changed the resume-time for the download then this would fix things (as this time it is in a paused state) but I can't see how it would work otherwise.

**Comment by ferdnyc** [comment]
...Huh. Sonofa— I guess not! I really thought it did. All of my testing for this report was with downloads that had been through plenty of other states, so I never noticed. I guess now I have to confront the fact that I'm so indecisive, I've _**never**_ just set a torrent to start at some future point and left it alone, without subsequently second-guessing myself and adjusting the start time, or starting and re-pausing it, or some other sort of fiddling.

You know, now that I say it like that, it actually seems totally plausible. Yeah, sounds like me.

**Comment by ferdnyc** [comment]
Confirming fixed, since 1.9.0.1_B18 stopped torrents can be started by setting a "Pause For..." timer and waiting for it to expire. Thanks!
